Basic UI, persist learner answers upon submit

itsjeyd opened the pull request:
cf. [OC-4159](https://tasks.opencraft.com/browse/OC-4159), [OC-4160](https://tasks.opencraft.com/browse/OC-4160)

This PR ships the following features:

- A basic UI for the LPD that is capable of rendering all types of questions that the MVP needs to support.
- Logic for storing answers to different types of questions in the DB.
- Logic for displaying saved answers back to the learner when revisiting the LPD.
- Functionality for expanding and collapsing individual sections of the learner profile.
- Logic that limits learner interactions with MRQs and Ranking questions according to the unique constraints of these question types.
    - MRQs don't allow learners to select more options than `MultipleChoiceQuestion.max_options_to_select`.
    - Rating questions don't allow learners to select the same rank for more than one option.

**Test instructions**

1. Run migrations.
1. Load test data for the `lpd` by running `./manage.py loaddata db-lpd.json` (see ticket attachments for the `db-lpd.json` file). If that doesn't work for some reason, launch the Django shell and `%cpaste` the commands from `commands.txt` (also attached to the ticket) to populate the DB.
1. Access the LPD in the LMS. (This assumes you've already [embedded](https://github.com/open-craft/learner-profile-dashboard/tree/27681e1a1379cee3b85b3135605129788e489d69#using-lpd-with-open-edx) the LPD as an LTI component in Open edX.)

1. Verify that you can collapse and expand individual sections of the LPD by clicking on their titles.

1. Refresh the page a few times. Observe that the order of options changes for questions that have `randomize_options` set to `True`. (When using test data from `db-lpd.json` or `commands.txt`, questions that are configured to display their options in random order include a multiple choice question ("Choose one of:") and a ranking question ("Rank these on a scale of 1 to 3:").) Observe that the order of options does **not** change for the remaining questions.

1. Answer a few questions and click "Submit your answers". Make sure to enter custom input for some of the answer options.

1. Refresh the page. Observe that the LPD displays the answers that you provided in the previous step, as well as any custom input you provided for specific answer options. This works for both regular refreshes and refreshes that [override the cache](https://support.mozilla.org/en-US/questions/1073264).

1. Verify that you can not select more than two options for the MRQ ("Choose at most two of:").
    - After selecting two options, checkboxes for the remaining (unselected) options will be disabled.
    - After deselecting one of the two selected options, all options will be enabled.
    - After refreshing the page with two options selected, checkboxes for the remaining (unselected) options will remain disabled.
    - After refreshing the page with a single option selected, all options will remain enabled.

1. Verify that you can not select the same rank for two different options belonging to the same ranking question. For example, if you select "1" for the "I love salad" option, selecting "1" for the "I love candy" option will result in "1" getting deselected for the "I love salad" option. The intended behavior for ranking questions is to never allow learners to rank more than `RatingQuestion.number_of_options_to_rank`, and for the value of each ranked option to be unique. (It is OK to rank *less* than `RatingQuestion.number_of_options_to_rank`, though.)

**Notes**

- This PR does not introduce custom CSS for the LPD. We will take care of that via a separate ticket.
- The build is currently failing due to a few leftover quality issues. To save time I will fix these after addressing review comments.
- The diff contains a few `print` and `console.log` statements that have been helpful for debugging purposes. I will remove these after addressing review comments.
- **Code review**: You can skip over third-party JS libraries added to `app/static/js/vendor/`.

**Reviewers**

- [x] @tomaszgy

tomaszgy commented:
@itsjeyd Just tested it, all works well! ;) Will go quickly to the review now.
One thing regarding no. 8 in testing instructions and refreshing the page: before refreshing the page, the user needs to submit the answers (by clicking _submit_ button), otherwise the answer isn't preserved. I guess that's intended behavior, right? :)

tomaszgy commented on lpd/views.py:
```diff
@@ -1,27 +1,124 @@
+                ),
+            )
+            if custom_input:
+                answer.custom_input = custom_input
```
Could this bit be moved above `update_or_create` call in the following form:
```python
defaults = dict(value=value)
if custom_input:
   defaults[custom_input] = custom_input
```

?

I think that way we could avoid hitting db another time, no?

itsjeyd replied:
@tomaszgy Definitely, that's a great idea. Applied that change in 82ab6fcd669707a6d83752961c2cf88f3119aea9.

tomaszgy commented on lpd/templates/question.html:
```diff
@@ -0,0 +1,117 @@
+                   class="custom-input"
+                   value="{{ answer_option_data.custom_input }}" />
+          {% endif %}
+          <input type="radio"
```
This is a matter of preference I guess, so please take this comment with a grain of salt, but I find it more intuitive when the checkbox is on the left side of the option ;)

itsjeyd replied:
@tomaszgy You're right, somehow that's more appealing visually (even without any custom CSS to style the questions)! Made that change: 6d8063066b8b223101089c18b9b8c90425e671fb.

itsjeyd commented on lpd/views.py:
```diff
@@ -1,27 +1,124 @@
+        try:
+            self._process_qualitative_answers(user, qualitative_answers)
+            self._process_quantitative_answers(user, quantitative_answers)
+        except Exception as e:  # pylint: disable=broad-except
```
@tomaszgy 👍 I enabled logging for `lpd.views` via 55dbc6a4bf128b4a0cbfe8241a13ab3548e38803.

itsjeyd replied:
@tomaszgy 👍 I enabled logging for `lpd.views` via 55dbc6a4bf128b4a0cbfe8241a13ab3548e38803.

itsjeyd commented:
@tomaszgy I addressed your comments and added tests. We don't have full coverage yet, so I'll create a follow-up ticket for addressing that.

Please have a look at the latest changes.

tomaszgy commented on lpd/tests/test_models.py:
```diff
@@ -1,11 +1,317 @@
+        """
+        Test string representation of `LikertScaleQuestion` model.
+        """
+        question = LikertScaleQuestionFactory(question_text='Is this a Likert scale question?')
```
I guess this is a nit but if we do:
```python
self.question_factory = LikertScaleQuestionFactory
```
 in the `setUp` method, then we could do here:
```python
question = self.question_factory(question_text='Is this a Likert scale question?')
```

(It's the same thing for the other test classes, I think)

tomaszgy commented on lpd/tests/test_models.py:
```diff
@@ -1,11 +1,317 @@
+        question = QualitativeQuestionFactory()
+        learner = UserFactory()
+        self.assertEqual(question.get_answer(learner), '')
+        answer = QualitativeAnswerFactory(learner=learner, question=question, text='This is not an answer.')
```
I see a René Magritte's fan here ;)

![Ceci n'est pas une pipe](https://upload.wikimedia.org/wikipedia/en/b/b9/MagrittePipe.jpg)

tomaszgy commented on lpd/tests/test_models.py:
```diff
@@ -1,11 +1,317 @@
+        question_numbers = sorted(random.sample(range(1, 100), 20))
+        for unused in range(5):
+            for question_factory in QUESTION_FACTORIES:
+                question = question_factory(section=section, number=question_numbers.pop(0))
```
Just to confirm: this will raise exception if we add another factory to `QUESTION_FACTORIES`
but that's intended, right?

Otherwise it might be better to exchange `20` in `random.sample()` for:

```python
5 * len(QUESTION_FACTORIES)
```

itsjeyd replied:
@tomaszgy It's not intended, I just don't expect that we will add more question types any time soon (so we won't need any additional `QUESTION_FACTORIES`) 🙂 

But it won't hurt to make this more robust in any case, so I've done that in 6187ffdf84edce67f833865819948493b8dd945c.

tomaszgy commented:
@itsjeyd 
Thanks, good job, I've added few small comments ;)

There's one thing I'm not completely sure - that is the multiple calls to `random` package's methods in `factories.py`. Wouldn't this:
1) Make fixing tests way harder (as we will not be sure what was the input dataset)? If so, I guess this could be solved be logging the input parameters somehow.
2) Hide test failures? (a PR could be merged as its CI builds didn't receive the failing input dataset; also, a failing dataset may appear later on in some other PR and could mislead us into thinking that this later PR is a root of the test failures)

So I think the question would be whether the calls to `random` affect (or will affect later on) unit tests. This I'm not sure so feel free to comment on this :)

Other than that, all looks good :)

itsjeyd commented:
@tomaszgy Thanks for the feedback! I addressed the remaining comments, please have another look.

Re: calls to `random`, you're right that these have the potential to opaque what's going on. What I've done for now is to add logging (4c53571c6a9267ed72cb71f6622d6b0e19ccd8f5) so that we can look up details about the data sets that a specific test run was using, if necessary. That was quicker than carefully crafting specific data sets. Also, I changed the logic of one of the tests so that it no longer needs to call `random` (f5b5b836327e697793cb7ed673d9930b59f84bb5). So now it's only 1 unit test out of 35 that is still using `random`. I think we can live with that, and we can always revisit this test when implementing additional features.

tomaszgy commented:
@itsjeyd Thanks, looking good now! 👍 

For the added testing logger - would be nice the logs are being kept in CircleCI if that's easy to set-up (because then we can easily reproduce failing input data). This could be done in another future ticket though.

Having said all of that:
:+1: LGTM

- [x] I tested this:
1) I followed the testing instructions.
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

itsjeyd commented:
Thanks @tomaszgy! Merging now.
